### PR TITLE
SONARJAVA-2391 Ignore thread which use runnables

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/ThreadOverridesRunCheck.java [hunk 1]
```diff
@@ -72,4 +87,45 @@ private static boolean overridesRunMethod(Symbol.TypeSymbol classSymbol) {
...
+  private static class SuperRunnableVisitor extends BaseTreeVisitor {
+
+    private boolean callSuperWithRunnable = false;
+
+    private static final MethodMatcher SUPER_THREAD = MethodMatcher.create().typeDefinition("java.lang.Thread").name("<init>").withAnyParameters();
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree tree) {
+      if (SUPER_THREAD.matches(tree) && ThreadOverridesRunCheck.hasRunnableArgument(tree.arguments())) {
+        callSuperWithRunnable = true;
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/ThreadOverridesRunCheck.java]: micro optimisation but : no need to visit further...

File: java-checks/src/main/java/org/sonar/java/checks/ThreadOverridesRunCheck.java [hunk 2]
```diff
@@ -72,4 +87,45 @@ private static boolean overridesRunMethod(Symbol.TypeSymbol classSymbol) {
...
+      }
+      super.visitMethodInvocation(tree);
+    }
+
+    @Override
+    public void visitClass(ClassTree tree) {
+      // skip inner classes
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/ThreadOverridesRunCheck.java]: Not sure I actually spotted the testcase where this skip of inner class is required.

Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/ThreadOverridesRunCheck.java]: There is one. [Line#47](https://github.com/SonarSource/sonar-java/pull/1589/files#diff-d93a2ec6a1ed7e239cc98988f8d389b5R47) would raise two issue if classes were not skipped.

Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/ThreadOverridesRunCheck.java]: modified my test case (see [Line#49](https://github.com/SonarSource/sonar-java/pull/1589/files#diff-d93a2ec6a1ed7e239cc98988f8d389b5R49))

**Pull request merged**